README.rst
.. image:: https://api.travis-ci.org/dataflake/dataflake.ldapconnection.svg?branch=master
   :target: https://travis-ci.org/dataflake/dataflake.ldapconnection

.. image:: https://readthedocs.org/projects/dataflakeldapconnection/badge/?version=latest
   :target: https://dataflakeldapconnection.readthedocs.io
   :alt: Documentation Status

.. image:: https://img.shields.io/pypi/v/dataflake.ldapconnection.svg
   :target: https://pypi.python.org/pypi/dataflake.ldapconnection
   :alt: PyPI

.. image:: https://img.shields.io/pypi/pyversions/dataflake.ldapconnection.svg
   :target: https://pypi.python.org/pypi/dataflake.ldapconnection
   :alt: Python versions

==========================
 dataflake.ldapconnection
==========================
This package provides an abstraction layer on top of ``pyldap``. It
offers a connection object with simplified methods for inserting, 
modifying, searching and deleting records in the LDAP directory tree.
Failover/redundancy can be achieved by supplying connection data for 
more than one LDAP server.


Documentation
=============
Full documentation is available at
http://dataflakeldapconnection.readthedocs.io

Bug tracker
===========
A bug tracker is available at
https://github.com/dataflake/dataflake.ldapconnection/issues


CHANGES.txt
Change log
==========

2.2 (unreleased)
----------------


2.1 (2018-06-29)
----------------
- test and declare support for Python 3.7


2.0 (2017-10-29)
----------------
- Python 3 compatibility
- Move from ``python-ldap`` to ``pyldap`` for Python 3 compatibility
- PEP 8 code cleanup
- added Travis CI automatic testing configuration
- use pkgutil-style namespace declaration
- package cleanup (``.gitignore``, ``MANIFEST.in``, ``README.rst``)
- docs cleanup (``Makefile``, ``conf.py``)
- tests cleanup (``tox.ini``, ``.travis.yml``)
- remove unsupported documentation bits
- fix coverage tests to only test this package
- remove coveralls from the Travis CI configuration
- make the character sets used for communicating with the LDAP libraries
  (``ldap_encoding``) and the calling side (``api_encoding``) 
  configurable in the :mod:`LDAPConnection` constructor.


1.5 (2012-04-20)
----------------
- Factored the ``dataflake.ldapconnection.tests.fakeldap`` module
  into a separate package ``dataflake.fakeldap``

- Restricted the supported Python versions to 2.6 and 2.7.

- removed old test_suite fossils


1.4 (2012-04-03)
----------------
- Added a simple tox configuration to quickly test with different 
  Python versions. As a result of successful testing, re-declare 
  compatibility with Python 2.4 and 2.5.

- Moved the documentation build dependencies into a setup extra so 
  building them without buildout becomes easier.

- Added setup.py aliases for creating a testing sandbox and the 
  documentation dependencies, and use them in the buildout 
  configuration.

- Extended ``fakeldap.RaisingFakeLDAPConnection`` to accept a list
  of exceptions to raise. On each call to the method that is set 
  to raise the exception, the first item in the exception list 
  is removed and raised. This allows testing code areas nested in 
  more than one ``try/except`` clause.


1.3 (2012-03-23)
----------------
- Update bootstrap.py to what ships with zc.buildout 1.5.2

- fakeldap: Add email characters and some non-ASCII characters to
  FLTR_RE, to be able to use them in searches. (Stefan Holek)

- fakeldap: Add unbind_s API. (Stefan Holek)

- fakeldap: Deep-copy entries before returning them. (Stefan Holek)

- fakeldap: Only return requested attributes. (Stefan Holek)

- fakeldap: Optionally disable password hashing. (Stefan Holek)

- fakeldap: Optionally maintain the 'memberOf' attribute of group
  members. (Stefan Holek)

- connection: Clean up encoding and decoding of values for when the
  api_encoding is set to None. (Stefan Holek)

- connection: New 'raw' kwarg for the search API. If true, search
  results are returned in the ldap_encoding. (Stefan Holek)

- connection: Store a connection's bind_dn and bind_pwd as is, and
  encode them before use. (Stefan Holek)

- connection: Fix a condition that caused rebinding to fail if only
  one of bind_dn and bind_pwd had changed. (Stefan Holek)

- connection: Add disconnect API. (Stefan Holek)

- connection: Allow to delete subsets from multi-valued attributes.
  (Stefan Holek)

- switched documentation to point to the new Git repository


1.2 (2010-08-09)
----------------
- Using ``id()`` is not random enough for a unique hash.


1.1 (2010-05-09)
----------------
- Updated Sphinx Makefile and configuration to be closer
  to the latest Sphinx version

- Greatly expand installation and testing documentation using
  ideas from Tres Seaver


1.0 (2010-04-12)
----------------
- Bug: fakeldap.FakeLDAPConnection wildcard searches did not 
  work correctly and returned too many matches.

- Bug: Improve behavior matching of standard python-ldap and 
  fakeldap by raising ldap.NO_SUCH_OBJECT where operations 
  target non-existing entries.

- Bug: Improve behavior matching of standard python-ldap and 
  fakeldap by raising ldap.ALREADY_EXISTS where operations 
  duplicate existing entries.

- Bug: Added tests for all fakeldap.FakeLDAPConnection methods
  and added tests for some other module classes and functions.

- Refactoring: Removed the fakeldap.initialize and explode_dn 
  functions. They were either not needed or needlessly 
  duplicating existing python-ldap features.

- Bug: python-ldap will no longer support the LDAP connection 
  class ldap.ldapobject.SmartLDAPObject with version 2.3.11.
  Replacing it with ReconnectLDAPObject.

- Bug: If a connection raised an LDAP exception inside 
  start_tls_s handling was broken.

- Feature: You can now add server definitions for servers that 
  support the StartTLS extended operation. Whereas the existing 
  secure connections using the ``ldaps`` protocol are encrypted 
  throughout, StartTLS is used through an unencrypted connection
  to request all further traffic to be encrypted.

- Refactoring: Switch tests to using the fakeldap LDAP connection 
  object wherever possible, and correct a few fakeldap and 
  LDAPConnection misbehaviors along the way.


1.0b1 (2010-02-01)
------------------

- Performing more rigorous input checking for DNs

- Made encoding/decoding more flexible by adding configuration 
  flags for the encoding used by the LDAP server and the 
  encoding for calls to and return values from the connection API.
  The default is backwards compatible (UTF-8 for the LDAP server 
  encoding, and Latin-15 for the API encoding).

- Factored the connection tests module into a series of modules,
  it was getting large and unwieldy.

- move the actual python-ldap connection from an attribute into
  a module-level cache since those connections cannot be
  pickled.

- Removed the rdn_attr attribute, which was used to try and 
  determine if a modify operation should trigger a modrdn. We 
  now fish the RDN attribute from the record's DN for this 
  purpose.

- Changed the way internal logging is done to avoid storing logger
  objects onto the connection instance unless it is explicitly

dataflake/ldapconnection/tests/base.py
##############################################################################
#
# Copyright (c) 2008-2012 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" unit tests base classes
"""

import six
import unittest

from dataflake.fakeldap import FakeLDAPConnection
from dataflake.fakeldap import FixedResultFakeLDAPConnection
from dataflake.fakeldap import RaisingFakeLDAPConnection
from dataflake.fakeldap.utils import hash_pwd


class LDAPConnectionTests(unittest.TestCase):

    def setUp(self):
        from dataflake.fakeldap import TREE

        super(LDAPConnectionTests, self).setUp()
        self.db = TREE
        # Put a record into the tree
        self.db.addTreeItems('dc=localhost')

    def tearDown(self):
        from dataflake.ldapconnection.connection import connection_cache

        super(LDAPConnectionTests, self).tearDown()
        self.db.clear()
        connection_cache.invalidate()

    def _getTargetClass(self):
        from dataflake.ldapconnection.connection import LDAPConnection
        return LDAPConnection

    def _makeOne(self, *args, **kw):
        conn = self._getTargetClass()(*args, **kw)
        return conn

    def _makeSimple(self):
        conn = self._makeOne('host', 636, 'ldap', FakeLDAPConnection)
        return conn

    def _makeRaising(self, raise_on, exc_class, exc_arg=None):
        ldap_connection = RaisingFakeLDAPConnection('conn_string')
        ldap_connection.setExceptionAndMethod(raise_on, exc_class, exc_arg)

        def factory(conn_string):
            ldap_connection.conn_string = conn_string
            return ldap_connection

        conn = self._makeOne('host', 389, 'ldaptls', factory)

        return conn, ldap_connection

    def _makeFixedResultConnection(self, results):
        ldap_connection = FixedResultFakeLDAPConnection()
        ldap_connection.search_results = results

        def factory(conn_string):
            ldap_connection.conn_string = conn_string
            return ldap_connection

        conn = self._makeOne('host', 389, 'ldaptls', factory)

        return conn

    def _factory(self, connection_string):
        of = FakeLDAPConnection(connection_string)
        return of

    def _addRecord(self, dn, **kw):
        record = self.db.addTreeItems(dn)
        for key, value in kw.items():
            if not isinstance(key, six.binary_type):
                key = key.encode('UTF-8')
            if key.lower() == b'userpassword':
                value = [hash_pwd(value)]
            elif not isinstance(value, list):
                value = [value]
            record[key] = value


dataflake/ldapconnection/tests/dummy.py
##############################################################################
#
# Copyright (c) 2008-2012 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" dummy: dummy test fixtures
"""

# From ISO-8859-1: Umlauts a, o, u and sharp s
UNENCODED_LATIN1 = u'\xe4\xf6\xfc\xdf'
# From ISO-8859-7 (Greek): Alpha, beta, gamma, delta
UNENCODED_GREEK = u'\u03b1\u03b2\u03b3\u03b4'


dataflake/ldapconnection/tests/fakeldap.py
##############################################################################
#
# Copyright (c) 2000-2012 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" A fake LDAP module for unit tests
"""

# BBB
from dataflake.fakeldap import FakeLDAPConnection
from dataflake.fakeldap import FixedResultFakeLDAPConnection
from dataflake.fakeldap import ldapobject
from dataflake.fakeldap import RaisingFakeLDAPConnection



dataflake/ldapconnection/ssha.py
##############################################################################
#
# Copyright (c) 2000-2008 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" SSHA encryption implementaion

$Id: SSHA.py 1485 2008-06-04 16:08:38Z jens $
"""

from binascii import a2b_base64
from binascii import b2a_base64
from binascii import Error as binascii_Error
from random import randrange
import sha


"""
SSHA is a modification of the SHA digest scheme with a salt
starting at byte 20 of the base64-encoded string.

This module contributed by Dirk Datzert
"""
# Source: [11]http://developer.netscape.com/docs/technote/ldap/pass_sha.html

def generate_salt():
    # Salt can be any length, but not more than about 37 characters
    # because of limitations of the binascii module.
    # 7 is what Netscape's example used and should be enough.
    # All 256 characters are available.
    salt = ''
    for n in range(7):
        salt += chr(randrange(256))

    return salt


def encrypt(password):
    password = str(password)
    salt = generate_salt()

    return b2a_base64(sha.new(password + salt).digest() + salt)[:-1]


def validate(reference, attempt):
    try:
        ref = a2b_base64(reference)
    except binascii_Error:
        # Not valid base64.
        return 0

    salt = ref[20:]
    compare = b2a_base64(sha.new(attempt + salt).digest() + salt)[:-1]

    return (compare == reference)



dataflake/ldapconnection/utils.py
##############################################################################
#
# Copyright (c) 2008-2012 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" Utility functions and constants
"""

import ldap
import six


BINARY_ATTRIBUTES = (b'objectguid', b'jpegphoto')


def escape_dn(dn, encoding='UTF-8'):
    """ Escape all characters that need escaping for a DN, see RFC 2253
    """
    if not dn:
        return dn

    escaped = ldap.dn.dn2str(ldap.dn.str2dn(dn))

    if isinstance(escaped, six.text_type):
        escaped = escaped.encode(encoding)

    return escaped


def dn2str(dn_parts, encoding='UTF-8'):
    """ Build a DN string from a parts structure
    """
    dn_list = []

    for dn_part in dn_parts:
        key = dn_part[0][0]
        value = dn_part[0][1]

        if not isinstance(key, six.binary_type):
            key = key.encode(encoding)

        if not isinstance(value, six.binary_type):
            value = value.encode(encoding)

        dn_list.append(b'%s=%s' % (key, value))

    return b','.join(dn_list)


dataflake/ldapconnection/connection.py
##############################################################################
#
# Copyright (c) 2008-2012 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" LDAPConnection: A class modeling a LDAP server connection

Instances of this class offer a simplified API to do searches, insertions,
deletions or modifications.
"""

import ldap
from ldap.dn import str2dn
from ldap.ldapobject import ReconnectLDAPObject
import ldapurl
import logging
from random import random
import six

from zope.interface import implementer

from dataflake.cache.simple import LockingSimpleCache
from dataflake.ldapconnection.interfaces import ILDAPConnection
from dataflake.ldapconnection.utils import BINARY_ATTRIBUTES
from dataflake.ldapconnection.utils import dn2str
from dataflake.ldapconnection.utils import escape_dn


default_logger = logging.getLogger('dataflake.ldapconnection')
connection_cache = LockingSimpleCache()
_marker = ()


@implementer(ILDAPConnection)
class LDAPConnection(object):
    """ LDAPConnection object

    See `interfaces.py` for interface documentation.
    """

    def __init__(self, host='', port=389, protocol='ldap',
                 c_factory=ReconnectLDAPObject, rdn_attr='',
                 bind_dn=b'', bind_pwd='', read_only=False, conn_timeout=-1,
                 op_timeout=-1, logger=None, ldap_encoding='UTF-8',
                 api_encoding='UTF-8'):
        """ LDAPConnection initialization

        Empty values for api_encoding or ldap_encoding mean "use unicode"
        """
        if isinstance(bind_dn, six.text_type) and ldap_encoding:
            bind_dn = bind_dn.encode(ldap_encoding)
        self.bind_dn = bind_dn
        self.bind_pwd = bind_pwd
        self.read_only = read_only
        self.c_factory = c_factory
        self._logger = logger
        self.ldap_encoding = ldap_encoding
        self.api_encoding = api_encoding
        self.hash = id(self) + random()

        self.servers = {}
        if host:
            self.addServer(host, port, protocol, conn_timeout, op_timeout)

    def logger(self):
        """ Get the logger
        """
        if self._logger is None:
            return default_logger

        return self._logger

    def addServer(self, host, port, protocol, conn_timeout=-1, op_timeout=-1):
        """ Add a server definition to the list of servers used
        """
        start_tls = False
        if protocol == 'ldaptls':
            protocol = 'ldap'
            start_tls = True
        hp = '%s:%s' % (host, port)
        conn = ldapurl.LDAPUrl(urlscheme=protocol, hostport=hp)
        server_url = conn.initializeUrl()
        self.servers[server_url] = {'url': server_url,
                                    'conn_timeout': conn_timeout,
                                    'op_timeout': op_timeout,
                                    'start_tls': start_tls}

    def removeServer(self, host, port, protocol):
        """ Remove a server definition from the list of servers used
        """
        hp = '%s:%s' % (host, port)
        conn = ldapurl.LDAPUrl(urlscheme=protocol, hostport=hp)
        server_url = conn.initializeUrl()
        if server_url in self.servers.keys():
            del self.servers[server_url]

    def connect(self, bind_dn=None, bind_pwd=None):
        """ initialize an ldap server connection

        This method returns an instance of the underlying `pyldap`
        connection class. It does not need to be called explicitly, all
        other operations call it implicitly.
        """
        if not self.servers:
            raise RuntimeError('No servers defined')

        if bind_dn is None:
            bind_dn = escape_dn(self._encode_incoming(self.bind_dn),
                                self.ldap_encoding)
            bind_pwd = self._encode_incoming(self.bind_pwd)
        else:
            bind_dn = escape_dn(self._encode_incoming(bind_dn),
                                self.ldap_encoding)
            bind_pwd = self._encode_incoming(bind_pwd)

        conn = self._getConnection()
        if conn is None:
            for server in self.servers.values():
                try:
                    conn = self._connect(server['url'],
                                         conn_timeout=server['conn_timeout'],
                                         op_timeout=server['op_timeout'])
                    if server.get('start_tls', None):
                        conn.start_tls_s()
                    break
                except (ldap.SERVER_DOWN, ldap.TIMEOUT, ldap.LOCAL_ERROR) as e:
                    conn = None
                    exc = e
                    continue

            if conn is None:
                msg = 'Failure connecting, last attempt: %s (%s)' % (
                            server['url'], str(exc) or 'no exception')
                self.logger().critical(msg, exc_info=1)

                if exc:
                    raise exc

            connection_cache.set(self.hash, conn)

        last_bind = getattr(conn, '_last_bind', None)
        if not last_bind or \
           last_bind[1][0] != bind_dn or \
           last_bind[1][1] != bind_pwd:
            conn.simple_bind_s(bind_dn, bind_pwd)

        return conn

    def _getConnection(self):
        """ Private helper to get my connection out of the cache
        """
        return connection_cache.get(self.hash)

    def _connect(self, connection_string, conn_timeout=5, op_timeout=-1):
        """ Factored out to allow usage by other pieces

        user_dn is assumed to have been encoded/escaped correctly
        """
        connection = self.c_factory(connection_string)

        # Deny auto-chasing of referrals to be safe, we handle them instead
        try:
            connection.set_option(ldap.OPT_REFERRALS, ldap.DEREF_NEVER)
        except ldap.LDAPError:  # Cannot set referrals, so do nothing
            pass

        # Set the connection timeout
        if conn_timeout > 0:
            connection.set_option(ldap.OPT_NETWORK_TIMEOUT, conn_timeout)

        # Set the operations timeout
        if op_timeout > 0:
            connection.timeout = op_timeout

        return connection

    def disconnect(self):
        """ Unbind the connection and invalidate the cache
        """
        conn = self._getConnection()
        if conn is not None:
            connection_cache.invalidate(self.hash)
            conn.unbind_s()

    def search(self, base, scope=ldap.SCOPE_SUBTREE, fltr='(objectClass=*)',
               attrs=None, convert_filter=True, bind_dn=None, bind_pwd=None,
               raw=False):
        """ Search for entries in the database
        """
        result = {'size': 0, 'results': [], 'exception': ''}
        if convert_filter:
            fltr = self._encode_incoming(fltr)
        base = escape_dn(self._encode_incoming(base),
                         self.ldap_encoding)
        connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)

        try:
            res = connection.search_s(base, scope, fltr, attrs)
        except ldap.PARTIAL_RESULTS:
            res_type, res = connection.result(all=0)
        except ldap.REFERRAL as e:
            connection = self._handle_referral(e)

            try:
                res = connection.search_s(base, scope, fltr, attrs)
            except ldap.PARTIAL_RESULTS:
                res_type, res = connection.result(all=0)

        for rec_dn, rec_dict in res:
            # When used against Active Directory, "rec_dict" may not be
            # be a dictionary in some cases (instead, it can be a list)
            # An example of a useless "res" entry that can be ignored
            # from AD is
            # (None, ['ldap://ForestDnsZones.PORTAL.LOCAL/DC=ForestDnsZones,\
            # DC=PORTAL,DC=LOCAL'])
            # This appears to be some sort of internal referral, but
            # we can't handle it, so we need to skip over it.
            try:
                items = list(rec_dict.items())
            except AttributeError:
                # 'items' not found on rec_dict
                continue

            if raw:
                rec_dict['dn'] = rec_dn
            else:
                for key, value in items:
                    if key == b'dn':
                        del rec_dict[key]
                    elif key.lower() not in BINARY_ATTRIBUTES:
                        if isinstance(value, (list, tuple)):
                            for i in range(len(value)):
                                value[i] = self._encode_outgoing(value[i])
                        else:
                            rec_dict[key] = self._encode_outgoing(value)

                rec_dict['dn'] = self._encode_outgoing(rec_dn)

            result['results'].append(rec_dict)
            result['size'] += 1

        return result

    def insert(self, base, rdn, attrs=None, bind_dn=None, bind_pwd=None):
        """ Insert a new record

        attrs is expected to be a mapping where the value may be a string
        or a sequence of strings.
        Multiple values may be expressed as a single string if the values
        are semicolon-delimited.
        Values can be marked as binary values, meaning they are not encoded
        as UTF-8, by appending ';binary' to the key.
        """
        self._complainIfReadOnly()
        base = escape_dn(self._encode_incoming(base), self.ldap_encoding)
        rdn = escape_dn(self._encode_incoming(rdn), self.ldap_encoding)

        dn = rdn + b',' + base
        attribute_list = []
        attrs = attrs and attrs or {}

        for attr_key, values in attrs.items():
            if attr_key.endswith(';binary'):
                is_binary = True
                attr_key = attr_key[:-7]
            else:
                is_binary = False

            if not isinstance(attr_key, six.binary_type):
                attr_key = self._encode_incoming(attr_key)

            if isinstance(values, six.string_types) and not is_binary:
                values = [x.strip() for x in values.split(';')]
            elif isinstance(values, six.binary_type) and not is_binary:
                values = [x.strip() for x in values.split(b';')]

            if values != ['']:
                if not is_binary:
                    values = [self._encode_incoming(x) for x in values]
                attribute_list.append((attr_key, values))

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)
            connection.add_s(dn, attribute_list)
        except ldap.REFERRAL as e:
            connection = self._handle_referral(e)
            connection.add_s(dn, attribute_list)

    def delete(self, dn, bind_dn=None, bind_pwd=None):
        """ Delete a record
        """
        self._complainIfReadOnly()

        dn = escape_dn(self._encode_incoming(dn), self.ldap_encoding)

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)
            connection.delete_s(dn)
        except ldap.REFERRAL as e:
            connection = self._handle_referral(e)
            connection.delete_s(dn)

    def modify(self, dn, mod_type=None, attrs=None, bind_dn=None,
               bind_pwd=None):
        """ Modify a record
        """
        self._complainIfReadOnly()

        unescaped_dn = self._encode_incoming(dn)
        dn = escape_dn(unescaped_dn, self.ldap_encoding)
        res = self.search(base=unescaped_dn, scope=ldap.SCOPE_BASE,
                          bind_dn=bind_dn, bind_pwd=bind_pwd, raw=True)
        attrs = attrs and attrs or {}
        cur_rec = res['results'][0]
        mod_list = []

        for key, values in list(attrs.items()):

            if key.endswith(';binary'):
                key = key[:-7]
                is_binary = True
            else:
                is_binary = False

            if not isinstance(key, six.binary_type):
                key = self._encode_incoming(key)

            if not is_binary:
                if isinstance(values, six.string_types):
                    values = [self._encode_incoming(x) for x in
                              values.split(';')]
                else:
                    values = [self._encode_incoming(x) for x in values]

            if isinstance(key, six.text_type):
                key = self._encode_incoming(key)

            if mod_type is None:
                if key not in cur_rec and values != [b'']:
                    mod_list.append((ldap.MOD_ADD, key, values))
                elif cur_rec.get(key, [b'']) != values and \
                        values not in ([b''], []):
                    mod_list.append((ldap.MOD_REPLACE, key, values))
                elif key in cur_rec and values in ([b''], []):
                    mod_list.append((ldap.MOD_DELETE, key, None))
            elif mod_type in (ldap.MOD_ADD, ldap.MOD_DELETE) and \
                    values == [b'']:
                continue
            elif mod_type == ldap.MOD_DELETE and \
                    set(values).difference(set(cur_rec.get(key, []))):
                continue
            else:
                mod_list.append((mod_type, key, values))

            attrs[key] = values

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)

            dn_parts = str2dn(dn)
            clean_dn_parts = []
            for dn_part in dn_parts:
                for (attr_name, attr_val, flag) in dn_part:
                    if isinstance(attr_name, six.text_type):
                        attr_name = self._encode_incoming(attr_name)
                    if isinstance(attr_val, six.text_type):
                        attr_val = self._encode_incoming(attr_val)
                    clean_dn_parts.append([(attr_name, attr_val, flag)])

            rdn_attr = clean_dn_parts[0][0][0]
            raw_rdn = attrs.get(rdn_attr, '')
            if isinstance(raw_rdn, six.string_types):
                raw_rdn = [raw_rdn]
            new_rdn = raw_rdn[0]

            if new_rdn:
                rdn_value = self._encode_incoming(new_rdn)
                if rdn_value != cur_rec.get(rdn_attr)[0]:
                    clean_dn_parts[0] = [(rdn_attr, rdn_value, 1)]
                    dn_parts[0] = [(rdn_attr, raw_rdn[0], 1)]
                    raw_utf8_rdn = rdn_attr + b'=' + rdn_value
                    new_rdn = escape_dn(raw_utf8_rdn, self.ldap_encoding)
                    connection.modrdn_s(dn, new_rdn)
                    dn = dn2str(clean_dn_parts)

            if mod_list:
                connection.modify_s(dn, mod_list)
            else:
                debug_msg = 'Nothing to modify: %s' % dn
                self.logger().debug(debug_msg)

        except ldap.REFERRAL as e:
            connection = self._handle_referral(e)
            connection.modify_s(dn, mod_list)

    def _handle_referral(self, exception):
        """ Handle a referral specified in the passed-in exception
        """
        payload = exception.args[0]
        info = payload.get('info')
        ldap_url = info[info.find('ldap'):]

        if ldapurl.isLDAPUrl(ldap_url):
            conn_str = ldapurl.LDAPUrl(ldap_url).initializeUrl()
            conn = self._connect(conn_str)
            conn.simple_bind_s(self._encode_incoming(self.bind_dn),
                               self._encode_incoming(self.bind_pwd))
            return conn
        else:
            raise ldap.CONNECT_ERROR('Bad referral "%s"' % str(exception))

    def _complainIfReadOnly(self):
        """ Raise RuntimeError if the connection is set to `read-only`

        This method should be called before any directory tree modfication
        """
        if self.read_only:
            raise RuntimeError(
                'Running in read-only mode, directory modifications disabled')

    def _encode_incoming(self, value):
        """ Encode a string value to the encoding set for the LDAP server

        - if "value" is unicode, it will be encoded to self.ldap_encoding, but
          only if self.ldap_encoding is set.
        - if "value" is not unicode, it is assumed to be encoded as
          self.api_encoding. It is decoded and encoded to self.ldap_encoding
          if self.ldap_encoding is set, unless self.api_encoding and
          self.ldap_encoding are identical. In that case the passed-in value
          is handed back unchanged.
        """
        if value is None:
            return None

        elif isinstance(value, six.text_type):
            if self.ldap_encoding:
                value = value.encode(self.ldap_encoding)

        else:
            if self.api_encoding != self.ldap_encoding:
                if self.api_encoding:
                    value = value.decode(self.api_encoding)

                    if self.ldap_encoding:
                        value = value.encode(self.ldap_encoding)

        return value

    def _encode_outgoing(self, value):
        """ Encode a string value to the API encoding

        - if "value" is unicode, it will be encoded to self.api_encoding, but
          only if self.api_encoding is set.
        - if "value" is not unicode, it is assumed to be encoded as
          self.ldap_encoding. It is decoded and encoded to self.api_encoding
          if self.api_encoding is set, unless self.ldap_encoding and
          self.api_encoding are identical. In that case the passed-in value
          is handed back unchanged.
        """
        if value is None:
            return None

        elif isinstance(value, six.text_type):
            if self.api_encoding:
                value = value.encode(self.api_encoding)

        else:
            if self.api_encoding != self.ldap_encoding:
                if self.ldap_encoding:
                    value = value.decode(self.ldap_encoding)

                    if self.api_encoding:
                        value = value.encode(self.api_encoding)

        return value
